gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java

```diff
@@ -113,6 +109,9 @@ public abstract class ClientContext {
   @Nullable
   public abstract String getQuotaProjectId();
 
+  /** Package-Private as this is to be shared to StubSettings */
+  abstract EndpointContext getEndpointContext();
```

Comment:
Pass the EndpointContext to the StubSettings. This is to set allow the StubSettings to be re-created with the original settings since `clientContext.getEndpoint()` returns the resolved endpoint and not the original config.

StubSettings will re-created the EndpointContext in the constructor and should be re-created with the settings users passed in.

Reply:
Alternative: We pass the EndpointContext to ApiCallContext. We can expose a new public getter to the ApiCallContext (i.e. `getEndpointContext()`), but that would add a public method.

Reply:
> Pass the created EndpointContext to the StubSettings. This is to set allow the StubSettings to be re-created with the original users settings

Really like it! No more public methods!